### ISSUE TITLE
callable-types: should allow a function type with generic be written as interface shape.

### ISSUE THREAD

**Issue by zheeeng** [palantir/tslint#4388]
**Rule: callable-types**
*An interface or literal type with just a call signature can be written as a function type.*

### Bug Report

- __TSLint version__: 5.11.0
- __TypeScript version__: 3.2.2
- __Running TSLint via__: VSCode

#### TypeScript code being linted

```ts
interface UseAlternativeToggle {
  // tslint:disable-next-line:callable-types
  <T1, T2 = T1 >(initial: T1, alternative: T2):
    { value: T1 | T2, toggle: () => void, toggleOn: () => void, toggleOff: () => void }
}
type UseBooleanToggle = (initial?: boolean) =>
    { value: boolean, toggle: () => void, toggleOn: () => void, toggleOff: () => void },

interface UseToggle extends UseAlternativeToggle, UseBooleanToggle {
}

const useAlternativeToggle: UseAlternativeToggle = <T1, T2 = T1>(initial: T1, alternative: T2) => {
  const [value, updateValue] = useState<T1 | T2>(initial)
  const toggle = useCallback(() => updateValue(prev => prev === initial ? alternative : initial), [])
  const toggleOn = useCallback(() => updateValue(alternative), [])
  const toggleOff = useCallback(() => updateValue(initial), [])

  return {
    value, toggle, toggleOn, toggleOff,
  }
}

const useBooleanToggle: UseBooleanToggle = (initial: boolean = false) => {
  const [value, updateValue] = useState(initial)
  const toggle = useCallback(() => updateValue(prev => !prev), [])
  const toggleOn = useCallback(() => updateValue(true), [])
  const toggleOff = useCallback(() => updateValue(false), [])

  return {
    value, toggle, toggleOn, toggleOff,
  }
}

function useToggle (...args: any[]) {
  if (args.length === 2) return useAlternativeToggle(args[0], args[1])

  return useBooleanToggle(args[0])
}
export default useToggle as UseToggle
```

with `tslint.json` configuration:

```json
{
    "rules": { "callable-types": true }
}
```

I'm writting the useToogle with overload signatures:
1. `UseAlternativeToggle` has generic. (I use interface, it's be suggested as to using type with function call signature)
2. `UseBooleanToggle` hasn't generic
3. `UseToogle` extends `UseAlternativeToggle`, `UseBooleanToggle`
4. function `useAlternativeToggle` as declared as `UseAlternativeToggle`
5. function `useBooleanToggle` as declared as `UseBooleanToggle`
6. function `useToggle` decide to use function `useAlternativeToggle` or `useBooleanToggle`
7. assert  function `useToggle` as `UseToogle`, and  function `useToggle` get two overload signatures.

But the problem is that  `UseAlternativeToggle` is be suggested to written as type shape by rule `callable-types`. 
```
<T1, T2 = T1 >(initial: T1, alternative: T2):
  { value: T1 | T2, toggle: () => void, toggleOn: () => void, toggleOff: () => void }
Interface has only a call signature — use `type UseAlternativeToggle = <T1, T2 = T1 >(initial: T1, alternative: T2) =>
    { value: T1 | T2, toggle: () => void, toggleOn: () => void, toggleOff: () => void }` instead.tslint(callable-types)
```

If we apply this rule, how can we leave the `UseToogle` without declaring any generic on it?

I expected after extending, `UseToogle` be
```ts
interface UseToogle {
   <T1, T2 = T1 >(initial: T1, alternative: T2):
    { value: T1 | T2, toggle: () => void, toggleOn: () => void, toggleOff: () => void }
  (initial?: boolean) =>
    { value: boolean, toggle: () => void, toggleOn: () => void, toggleOff: () => void }
}
```

**not**:
```ts
interface UseToogle<T1, T2> {
   (initial: T1, alternative: T2):
    { value: T1 | T2, toggle: () => void, toggleOn: () => void, toggleOff: () => void }
  (initial?: boolean) =>
    { value: boolean, toggle: () => void, toggleOn: () => void, toggleOff: () => void }
}
``` 

#### Actual behavior
```ts
interface UseAlternativeToggle {
  // tslint:disable-next-line:callable-types
  <T1, T2 = T1 >(initial: T1, alternative: T2):
    { value: T1 | T2, toggle: () => void, toggleOn: () => void, toggleOff: () => void }
}
```
is suggested be written as:
```ts
type UseAlternativeToggle<T1, T2 = T1 > = (initial: T1, alternative: T2) =>
    { value: T1 | T2, toggle: () => void, toggleOn: () => void, toggleOff: () => void }
```

#### Expected behavior
Tough `UseAlternativeToggle` has only one function call signature, it should be perserved as its original type.

I think:
```ts
interface SomeFunction {
    <T>(i: T) => T
}
```
and
```ts
type SomeFunction<T> = (i: T) => T
```

have different usages, so they are not equivalent.

**Comment by zheeeng** [comment]
@JoshuaKGoldberg I made a mistake. I just found that

```ts
interface UseAlternativeToggle {
  <T1, T2 = T1 >(initial: T1, alternative: T2):
    { value: T1 | T2, toggle: () => void, toggleOn: () => void, toggleOff: () => void }
}
type UseBooleanToggle = (initial?: boolean) =>
    { value: boolean, toggle: () => void, toggleOn: () => void, toggleOff: () => void },

interface UseToggle extends UseAlternativeToggle, UseBooleanToggle {
}
````

and 

```ts
type UseAlternativeToggle<T1, T2 = T1 > = (initial: T1, alternative: T2) =>
    { value: T1 | T2, toggle: () => void, toggleOn: () => void, toggleOff: () => void }
type UseBooleanToggle = (initial?: boolean) =>
    { value: boolean, toggle: () => void, toggleOn: () => void, toggleOff: () => void },

interface UseToggle extends UseAlternativeToggle, UseBooleanToggle {
}
````

are equivalent. And:

```ts
interface UseToggle extends UseAlternativeToggle, UseBooleanToggle {}
```

and

```ts
type UseToggle = UseAlternativeToggle & UseBooleanToggle
```

are equivalent.

I should close this issue.

In my VS Code,  `callable-types --fix` doesn't effect on
```ts
interface UseToggle extends UseAlternativeToggle, UseBooleanToggle {
}
```
It only auto fix `UseAlternativeToggle `.
What else I can feedback to you for the auto fixing?

(unrelated: React hook i'm lovin it)

**Comment by JoshuaKGoldberg** [comment]
Great, glad this is working as expected!

> `callable-types --fix` doesn't effect on ...`

Yeah, that looks like a bug. You can file a new issue if you'd like, it seems like it might be difficult to fix but seems reasonable.
